### PR TITLE
Add disk caching for GlyphStores to reduce LOH usage

### PR DESCRIPTION
**Pull request by peppy**
Reduces memory spiking caused by `TimedExpiryCache` keeping cached raw glyph atlas data in memory for 5 seconds at a time.

In osu!'s case, on desktop (left is new method, right is old):

image

and the same scenario on mobile (benefitting from `ModeratePooling` configuration):

image

Performance is slightly worse off, but should be made back many times over by the fact that after the first cache, PNG decoding is no longer required.

To confirm we aren't much worse off, I have a [benchmark branch](https://github.com/peppy/osu-framework/tree/benchmark) running a relatively intensive scenario (time is for retrieving 10 glyphs, plus loading the full font)

```
|                           Method |     Mean |     Error |    StdDev |   Median |    Gen 0 |    Gen 1 |   Gen 2 |  Allocated |
|--------------------------------- |---------:|----------:|----------:|---------:|---------:|---------:|--------:|-----------:|
|        BenchmarkTimedExpiryCache | 17.35 ms | 0.0905 ms | 0.0802 ms | 17.32 ms |  93.7500 |  62.5000 | 62.5000 | 4170.28 KB |
| BenchmarkTimedExpiryCacheTrimmed | 14.13 ms | 0.0292 ms | 0.0244 ms | 14.13 ms | 125.0000 | 109.3750 | 93.7500 | 4170.29 KB |

|                           Method |     Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |---------:|----------:|----------:|--------:|------:|------:|----------:|
|        BenchmarkRawSaveNoCaching | 19.16 ms | 0.0540 ms | 0.0505 ms | 93.7500 |     - |     - | 475.65 KB |
| BenchmarkRawSaveNoCachingTrimmed | 16.64 ms | 0.0856 ms | 0.0715 ms | 93.7500 |     - |     - | 475.68 KB |
```

I believe there is still room for improvements as far as the remaining allocation overhead goes, but this is already miles better.

- This should also hopefully resolve https://github.com/ppy/osu/issues/3277, which looks to be caused by the `TimedExpiryCache` expiring before a load operation finished.

- This also adds thread safety to font lookups (at a `FontStore` level).

### CODE REVIEW

**Comment by peppy** [comment]
I guess the potential alternative for this would be to create a packing tool which did this at build time. We may want to consider this in the future.

**Comment by smoogipoo** [comment]
image

Seems to be copying more than required.

File: osu.Framework/IO/Stores/FontStore.cs [hunk 1]
```diff
@@ -106,18 +117,23 @@ public override void RemoveStore(IResourceStore<TextureUpload> store)
...
         }
 
+        private readonly object fontLookupLock = new object();
+
         public override Texture Get(string name)
         {
             var found = base.Get(name);
 
-            if (found == null)
+            lock (fontLookupLock)
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/IO/Stores/FontStore.cs]: this needs to also be around the line above, i think. it's meant to ensure the actual upload can only occur once per request

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/IO/Stores/FontStore.cs]: but the `base` one is locked

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/IO/Stores/FontStore.cs]: The idea is to block the lookup/store portion too, else two threads could get to this point at the same time and cause a double texture upload.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/IO/Stores/FontStore.cs]: Never mind.

**Pull request merged**